File: c++/Arrays/QuickSort.cpp
#include<iostream>
#include<vector>

using namespace std;
class Solution{
    public:
    void quickSort(vector<int>&arr, int low, int high){
        if(low<high){
            int pivot = partition(arr,low,high);
            quickSort(arr,low,pivot-1);
            quickSort(arr,pivot+1,high);
        }
    }
    int partition(vector<int>&arr,int low,int high){
        int pivot = arr[high];
        int i = low - 1;
        for(int j = low;j<high;j++){
            if(arr[j]<pivot){
                i++;
                swap(arr[i],arr[j]);
            }
        }
        swap(arr[i+1],arr[high]);
        return i+1;
    }
};
int main() {
    Solution solution;
    vector<int> arr = {10, 7, 8, 9, 1, 5}; // Example input
    int n = arr.size();
    solution.quickSort(arr, 0, n - 1); // Call the function
    cout << "Sorted array: ";
    for (int num : arr) {
        cout << num << " "; // Output the sorted array
    }
    cout << endl; // New line for better readability
    return 0; // Return success
}

File: c++/sorting/quickSort.java
public class quickSort {
    public static void main(String[] args) {
        int[] arr = {10, 7, 8, 9, 1, 5};
        int n = arr.length;

        quicksort(arr, 0, n - 1);
        System.out.println("Sorted array: ");
        printArray(arr);
    }

    static void quicksort(int arr[], int low, int high) {
        if (low < high) {
            // pi is partitioning index, arr[pi] is now at right place
            int pivotIndex = partition(arr, low, high);
            quicksort(arr, low, pivotIndex - 1); // Before pi
            quicksort(arr, pivotIndex + 1, high); // After pi
        }
    }

    static int partition(int arr[], int low, int high) {
        int pivot = arr[high]; // pivot
        int i = (low - 1); // Index of smaller element

        for (int j = low; j < high; j++) {
            // If current element is smaller than or equal to pivot
            if (arr[j] <= pivot) {
                i++;
                // swap arr[i] and arr[j]
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        // swap arr[i + 1] and arr[high] (or pivot)
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;
    }

    static void printArray(int arr[]) {
        for (int value : arr) {
            System.out.print(value + " ");
        }
        System.out.println();
    }
}
